### DISCUSSION TITLE
Set "Virtual" where "Downloadable"

### DISCUSSION THREAD

**Discussion by ZorioLan** [woocommerce/woocommerce#41849] [General]
Hello. We use physical and virtual goods. There are a lot of goods, so you need to solve this problem, for example, through sql. For virtual goods, we used the "Downloadable" checkbox, without "Virtual". After the update, the downloaded products began to require delivery, although the products themselves indicate that delivery is not needed. The question is, how can I add "Virtual" to all "downloadable " (non-physical) products? Is there any way to do this via sql?

**Comment by mattsherman** [comment]
Hi @ZorioLan!

Is this using the [new product editor](https://developer.woo.com/2023/11/17/getting-to-know-woo-extensibility-in-the-new-product-editor/), or the classic product editor experience?

What version of WooCommerce are you using? What version of WooCommerce worked differently?

From my experience, the downloadable and virtual traits have always been independent... because you can have physical products (that require delivery) that also have downloadable files. And you can have virtual products that aren't downloadable.

I'd like to understand if there is something that changed.

**Reply by ZorioLan** [reply]
Hi! Woocommerce Version 7.9.0.We are still using the Dokan plugin. After updating Woocommerce to version 8.3.1, we had to bring back version 7.9.0 and after that, many products that are marked as "Downloadable" began to require delivery. Our sellers have a lot of products and editing each product is long and difficult. We are looking for a solution that can make a massive update of products, namely, products that are marked as "Downloadable", mark them also as "Virtual". Can you suggest an sql query that can do this?

**Reply by mattsherman** [reply]
Thanks for the additional information!

Please note that this discussion group is not an official support channel for WooCommerce. For official support, I'd suggest you please visit [our support page](https://woo.com/my-account/contact-support/).

I'm unfamiliar with the Dokan plugin, and if you are having issues related to that plugin, I'd suggest you reach out to the creator of that plugin for help. It is possible that a recent change in WooCommerce has affected how that plugin works.

That said, let me try to provide some information here that may help you!

Running SQL statements without a backup of the database and a complete understanding of the data structure can be harmful. As I am not 100% positive of the state of your system or any configuration options/plugins that could affect things, I don't feel confident providing a SQL statement to update the values. However, I will share this SQL statement that queries and returns the current values of the traits for all downloadable products. If you are familiar with SQL this may help you to craft a query to make the update (please, backup your DB and fully test any such query on a backup/test site before running it on a production system!).

```sql
SELECT
	p.ID,
	p.post_title,
	pm_downloadable.meta_value AS downloadable_trait,
	pm_virtual.meta_value AS virtual_trait
FROM
	wp_posts AS p
	INNER JOIN wp_postmeta AS pm_downloadable ON p.ID = pm_downloadable.post_id
	LEFT JOIN wp_postmeta AS pm_virtual ON p.ID = pm_virtual.post_id
		AND pm_virtual.meta_key = '_virtual'
WHERE
	p.post_type = 'product'
	AND p.post_status = 'publish'
	AND pm_downloadable.meta_key = '_downloadable'
	AND pm_downloadable.meta_value = 'yes';
```
